### PR TITLE
Additional function arg unpacking (PEP 448)

### PR DESCRIPTION
**Pull request by dlech**
This is an attempt to implement PEP 448 for function arguments. It allows multiple *args and **args to be unpacked. (Popular request in #1329).

I'm curious to see how it affects code size and performance.

There are still some todo items:
- There is still a null object emitted in the bytecodes that used to be the **args that needs to be removed
- There are probably edge cases where memory allocation is wrong, especially in the case where one or more of the *args are iterators instead of list/tuple.

### CODE REVIEW

**Comment by tve** [comment]
might want to fix the typo in the title ;-)

**Comment by dlech** [comment]
If that is my worst mistake, we are in good shape 😆

**Comment by dpgeorge** [comment]
Thanks for working on this.  It's a tough problem, especially to make it efficient (in RAM usage, execution time, code size).  Properly reviewing this would take some time.

For the record, the code size change is:
```
   bare-arm:   +48 +0.072% 
minimal x86:  +128 +0.086% 
   unix x64:  +144 +0.029% 
unix nanbox:  +272 +0.061% 
      stm32:   +52 +0.013% PYBV10
     cc3200:   +48 +0.026% 
    esp8266:   +72 +0.011% GENERIC
      esp32:   +36 +0.003% GENERIC[incl -32(data)]
        nrf:   +52 +0.036% pca10040
       samd:   +48 +0.047% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```

**Comment by dpgeorge** [comment]
Can you please rebase this on latest master?

**Comment by dlech** [comment]
Rebased.

**Comment by dlech** [comment]
> * There is still a null object emitted in the bytecodes that used to be the **args that needs to be removed

Added a commit to address this todo item.

File: py/compile.c [hunk 1]
```diff
@@ -2397,24 +2397,25 @@ STATIC void compile_trailer_paren_helper(compiler_t *comp, mp_parse_node_t pn_ar
...
             if (MP_PARSE_NODE_STRUCT_KIND(pns_arg) == PN_arglist_star) {
-                if (star_flags & MP_EMIT_STAR_FLAG_SINGLE) {
-                    compile_syntax_error(comp, (mp_parse_node_t)pns_arg, MP_ERROR_TEXT("can't have multiple *x"));
+                if (star_flags & MP_EMIT_STAR_FLAG_DOUBLE) {
+                    compile_syntax_error(comp, (mp_parse_node_t)pns_arg, MP_ERROR_TEXT("* arg after **"));
                     return;
                 }
                 star_flags |= MP_EMIT_STAR_FLAG_SINGLE;
-                star_args_node = pns_arg;
+                star_args |= 1 << i;
```
Comment (dpgeorge) [inline py/compile.c:2410]: It'd probably be a good idea to check that this still fits in a small int, and raise an exception if not.

File: tests/basics/python34.py [hunk 2]
```diff
@@ -6,27 +6,23 @@
...
 
-# from basics/fun_kwvarargs.py
-# test evaluation order of arguments (in 3.4 it's backwards, 3.5 it's fixed)
-def f4(*vargs, **kwargs):
-    print(vargs, kwargs)
+
 def print_ret(x):
     print(x)
     return x
-f4(*print_ret(['a', 'b']), kw_arg=print_ret(None))
```
Comment (dpgeorge) [inline tests/basics/python34.py:16]: I think this test should be retained, it's a good test for evaluation order.  Maybe put it back in `basics/fun_kwvarargs.py` now that both uPy and CPy agree on the output?

**Comment by dpgeorge** [comment]
>> There is still a null object emitted in the bytecodes that used to be the **args that needs to be removed
>
> Added a commit to address this todo item.

IMO it makes sense to squash this commit into the first one, so that the `cmd_showbc` test does not change so much and then get reverted.  That would make it a bit clearer as to the impact of this change, and how it changes generated bytecode.

> There are probably edge cases where memory allocation is wrong, especially in the case where one or more of the *args are iterators instead of list/tuple.

Do you mean it might segfault in an edge case?  Or that it simply makes a bad guess at how much memory to preallocate?  As far as I can tell it all looks OK.

-----------

In fact this has exposed a bug in current master, the following will segfault:
```
def f(*a, **b):
    print(a)
    print(b)
f(*range(100), **{str(i):i for i in range(100)})
```
With this PR, that bug is fixed.

**Comment by dpgeorge** [comment]
I think the general approach here is very clever.  When `*seq` or `**dict` is used as an argument the Python stack looks like this:
```
fun(arg0, arg1, *seq, key0=value0, key1=value1, **dict)
old: fun arg0 arg1 key0 value0 key1 value1 seq dict
new: fun arg0 arg1 seq key0 value0 key1 value1 NULL dict 4
```
For simple cases:
```
fun(*seq)
old: fun seq NULL
new: fun seq 1
```
```
fun(**dict)
old: fun NULL dict
new: fun NULL dict 0
```
Also in the case of using `**dict` the bytecode for the opcode increased by 1 byte because in now needs 2 bytes to encode the `unum = (npos | nkw <<8)` value.  As such `fun(**dict)` uses 2 extra bytes in total in the bytecode (one for the unum, one for the seq bitmap).

I think this can be optimised.  Eg if npos==0 then don't store the bitmap.  And if npos==0 and nkw==0 then implicitly add 1 to nkw.

**Comment by dlech** [comment]
> IMO it makes sense to squash this commit into the first one,

Agreed.

> Do you mean it might segfault in an edge case? Or that it simply makes a bad guess at how much memory to preallocate?

The latter.

> I think this can be optimised. Eg if npos==0 then don't store the bitmap. And if npos==0 and nkw==0 then implicitly add 1 to nkw.

I will have a look.

**Comment by dlech** [comment]
I have updated this with the suggested changes. With regard to the suggested optimizations, I did it a bit differently. " if npos==0 then don't store the bitmap." was simple enough but "And if npos==0 and nkw==0 then implicitly add 1 to nkw." got messy since it requires checking for this in many places. So, I figured that if we were going to all all of these checks anyway, we could do it slightly differently and optimize (at least in terms of fewer bytes emitted) more cases. See the commit message for the full explanation.

File: py/compile.c (hunk 1 above)
Comment (dlech) [reply py/compile.c:2422]: I'm not sure this was worth the extra code size since it is rare that anyone would use that many arguments, but I added it anyway.

File: tests/basics/python34.py (hunk 2 above)
Comment (dlech) [reply tests/basics/python34.py:16]: done.

**Comment by dpgeorge** [comment]
> Given that arg unpacking is probably not frequently used, I'm not sure the trade off in compiler size/complexity is worth the savings in emitted byte code size.

You might be right, it does look complicated.  I will review it.

**Comment by peterhinch** [comment]
FWIW I think this is a very useful feature - subject to code size, of course :)

For example in GUI code, widgets may have multiple constructor args to define colors, shapes, fonts, sizes, callbacks etc. A `dict` enables a set of common attributes to be assigned to multiple widgets. The ability to assign more than one `dict` would be useful. You might have one to define attributes common to all controls and another for additional attributes applicable only to (say) `Meter` instances.

File: tests/basics/fun_kwargs.py [hunk 3]
```diff
@@ -33,3 +33,15 @@ def f3(a, b, *args):
...
+def f4(*vargs, **kwargs):
+    print(vargs, kwargs)
+
+
+def print_ret(x):
+    print(x)
+    return x
+
+
+f4(*print_ret(["a", "b"]), kw_arg=print_ret(None))
```
Comment (dpgeorge) [inline tests/basics/fun_kwargs.py:47]: this test should go in `tests/baics/fun_kwvarargs.py`

**Comment by dpgeorge** [comment]
Code size change for supporting multiple `*` and `**` args is:
```
   bare-arm:   +80 +0.139% 
minimal x86:  +133 +0.081% 
   unix x64:  +144 +0.028% 
unix nanbox:  +300 +0.065% 
      stm32:   +68 +0.017% PYBV10
     cc3200:   +80 +0.044% 
    esp8266:   +92 +0.013% GENERIC
      esp32:   +92 +0.006% GENERIC[incl -16(data)]
        nrf:   +16 +0.009% pca10040
        rp2:   +48 +0.010% PICO
       samd:   +64 +0.046% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```
Then the additional code size change on top of that for the encoding optimisation is:
```
   bare-arm:  +120 +0.208% 
minimal x86:  +204 +0.124% 
   unix x64:  +374 +0.071% 
unix nanbox:  +592 +0.128% 
      stm32:  +160 +0.041% PYBV10
     cc3200:  +128 +0.070% 
    esp8266:  +156 +0.022% GENERIC
      esp32:  +188 +0.012% GENERIC
        nrf:  +148 +0.085% pca10040
        rp2:  +136 +0.028% PICO
       samd:   -56 -0.040% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```
I'd say that optimisation is not worth it.  It's relatively rare to use `*` / `**` args and there are better ways than that to spend bytes on reducing/optimising the bytecode.

So please remove that encoding optimisation commit from this PR.  That should also fix the CI issues with native emitter.

File: py/vm.c [hunk 4]
```diff
@@ -927,8 +927,8 @@ unwind_jump:;
                     // unum & 0xff == n_positional
                     // (unum >> 8) & 0xff == n_keyword
                     // We have following stack layout here:
-                    // fun arg0 arg1 ... kw0 val0 kw1 val1 ... seq dict <- TOS
-                    sp -= (unum & 0xff) + ((unum >> 7) & 0x1fe) + 2;
+                    // fun arg0 arg1 ... kw0 val0 kw1 val1 ... bitmap <- TOS
```
Comment (dpgeorge) [inline py/vm.c]: In this commit the `bitmap` does not exist so it should be `seq` instead.  I.e., this commit should just delete the `dict` entry.  Then in the next commit, `seq` is renamed to `bitmap`.

**Comment by dpgeorge** [comment]
Performance change (without the encoding optimisation commit), on PYBv1.0:
```
diff of scores (higher is better)
N=100 M=100                        p0 ->         p1         diff      diff% (error%)
bm_chaos.py                    351.40 ->     355.38 :      +3.98 =  +1.133% (+/-0.01%)
bm_fannkuch.py                  77.31 ->      77.98 :      +0.67 =  +0.867% (+/-0.01%)
bm_fft.py                     2526.30 ->    2549.84 :     +23.54 =  +0.932% (+/-0.00%)
bm_float.py                   5905.82 ->    5939.79 :     +33.97 =  +0.575% (+/-0.00%)
bm_hexiom.py                    47.17 ->      47.27 :      +0.10 =  +0.212% (+/-0.01%)
bm_nqueens.py                 4419.24 ->    4451.92 :     +32.68 =  +0.739% (+/-0.00%)
bm_pidigits.py                 638.01 ->     640.70 :      +2.69 =  +0.422% (+/-0.23%)
core_import_mpy_multi.py       553.34 ->     556.64 :      +3.30 =  +0.596% (+/-0.01%)
core_import_mpy_single.py       68.36 ->      68.69 :      +0.33 =  +0.483% (+/-0.01%)
core_qstr.py                    63.73 ->      63.86 :      +0.13 =  +0.204% (+/-0.01%)
core_yield_from.py             356.12 ->     355.00 :      -1.12 =  -0.315% (+/-0.00%)
misc_aes.py                    411.90 ->     413.86 :      +1.96 =  +0.476% (+/-0.01%)
misc_mandel.py                3350.85 ->    3414.97 :     +64.12 =  +1.914% (+/-0.00%)
misc_pystone.py               2225.86 ->    2232.44 :      +6.58 =  +0.296% (+/-0.00%)
misc_raytrace.py               368.11 ->     372.18 :      +4.07 =  +1.106% (+/-0.00%)
viper_call0.py                 572.83 ->     572.73 :      -0.10 =  -0.017% (+/-0.00%)
viper_call1a.py                546.71 ->     546.71 :      +0.00 =  +0.000% (+/-0.00%)
viper_call1b.py                436.01 ->     435.91 :      -0.10 =  -0.023% (+/-0.01%)
viper_call1c.py                440.59 ->     440.51 :      -0.08 =  -0.018% (+/-0.00%)
viper_call2a.py                532.82 ->     532.85 :      +0.03 =  +0.006% (+/-0.01%)
viper_call2b.py                377.24 ->     377.17 :      -0.07 =  -0.019% (+/-0.00%)

```

File: py/vm.c (hunk 4 above)
Comment (dlech) [reply py/vm.c]: done

File: tests/basics/fun_kwargs.py (hunk 3 above)
Comment (dlech) [reply tests/basics/fun_kwargs.py]: done

**Comment by dlech** [comment]
The changes in code coverage show some potential issues: 

- [this branch](https://app.codecov.io/gh/micropython/micropython/compare/5807/diff#D3L2418) is not hit because the cpydiff isn't run in the coverage tests.
- [this branch](https://app.codecov.io/gh/micropython/micropython/compare/5807/diff#D4L816) and [this branch](https://app.codecov.io/gh/micropython/micropython/compare/5807/diff#D4L862) is no longer hit because we are now allocating one `mp_obj_t` for each kw arg, including the ** args plus the size of the ** arg if available. So we are allocating one extra `mp_obj_t` that isn't being used for each ** arg.

**Comment by dlech** [comment]
I have added a commit to address the overallocation issue. But it doesn't seem to affect the code coverage, so there is probably something else I am missing.

**Comment by dlech** [comment]
I've added a few more commits that should take care of the dropped coverage issues. So the only question that remains is should we add a test that duplicates the cpydiff to get the last bit of coverage or should we just include all cpydiff in the coverage run?

**Comment by dpgeorge** [comment]
> So the only question that remains is should we add a test that duplicates the cpydiff to get the last bit of coverage

Yes I think we should add another test for this in the main test suite (this is to check when there are too many positional args, right?).  Because cpydiff tests are intended to be changed/removed if/when the feature becomes compatible with CPython.  Thus cpydiff tests cannot be relied upon for coverage.

File: docs/differences/python_35.rst [hunk 5]
```diff
@@ -8,7 +8,7 @@ Below is a list of finalised/accepted PEPs for Python 3.5 grouped into their imp
    +----------------------------------------------------------------------------------------------------------+---------------+
    | **Extensions to the syntax:**                                                                            | **Status**    |
    +--------------------------------------------------------+-------------------------------------------------+---------------+
-   | `PEP 448 <https://www.python.org/dev/peps/pep-0448/>`_ | additional unpacking generalizations            |               |
+   | `PEP 448 <https://www.python.org/dev/peps/pep-0448/>`_ | additional unpacking generalizations            | Completed     |
```
Comment (dpgeorge) [inline docs/differences/python_35.rst:11]: I don't think PEP 0448 is fully complete, we still don't support multiples in tuples, dicts etc.  I don't know if we'll ever do that.

I suggest putting "Partially done" here.

File: py/runtime.c [hunk 6]
```diff
@@ -715,16 +714,37 @@ void mp_call_prepare_args_n_kw_var(bool have_self, size_t n_args_n_kw, const mp_
     uint args2_alloc;
     uint args2_len = 0;
 
+    // Try to get a hint for unpacked * args length
+    int list_len = 0;
```
Comment (dpgeorge) [inline py/runtime.c:718]: Now that this and `kw_dict_len` are signed, it's not clear to me that the calculations that involve these variables are guaranteed to be >= 0.... did you follow through all cases to check this?

Comment (dlech) [reply py/runtime.c:718]: They are guaranteed to be >= 0. `list_len` can be -1 if and only if `n_args` > 0 and it is only used as `n_args + list_len`. Likewise, `kw_dict_len` can be -1 if and only if `n_kw` is > 0 and is only used as `n_kw + kw_dict_len`. Neither value can ever by less than -1 since 1 is subtracted from an unsigned value.

We could add a couple of `assert()`s to try to catch any future regressions if the code is changed.

Comment (dpgeorge) [reply py/runtime.c:718]: `list_len` could be -2 if there are two lists, right?  Or even more negative.  But it sounds like that will still be offset by `n_args`.

Same with `kw_dict_len`?

Maybe there needs to be a test that passes in a few empty lists and empty dicts?

Comment (dlech) [reply py/runtime.c:718]: > Neither value can ever by less than -1 since 1 is subtracted from an unsigned value.

This statement is not accurate. You are correct. The -1 is in a for loop that loops `n_args` or `n_kw` times, so `list_len` and `kw_dict_len` can be at the smallest `-n_args` and `-kw_dict_len` respectively. But is is still mathematically provable that `n_args + list_len` and `n_kw + kw_dict_len` can never be < 0.

Comment (dpgeorge) [reply py/runtime.c:718]: I think the `int` should be `ssize_t`

**Comment by dpgeorge** [comment]
Thanks for the updates.  This is looking very good now.  Merged in 1e99d29f362f8cccc60a36fcbd8590404c719f40 through 2e3f2045f9a7511987e69a8be0d3d117502e4bb1

**Comment by dpgeorge** [comment]
I have a few extra coverage tests to add in a separate PR.
